### PR TITLE
Site Settings: Use Redux site for siteSettings route and settings main component

### PR DESCRIPTION
**Pull request by oskosk**
This PR updates the `siteSettings` route in the settings controller and the main controller component to not use `sites-list` for retrieving the current site's settings.

#### Changes introduced by this PR

* Makes the site settings controller drop usage of `sites.once` for the `siteSettings` route. (Involving general, writing, discussion, traffic, and security tabs).
* Drops the passing down ofthe `sites` object to `SiteSettingsComponent` from the controller.
* Refactors `SiteSettingsComponent` to pick the current site from the Redux state.
* ES6ifies the `SiteSettingsNavigation` component and makes it use the selected site form the redux state.
* Makes the DefaultPostForm component only query post formats if the site settings have been loaded.
* Makes the `SectionSecurity` component pick the selected site from Redux state.
* Makes the `PressThis` component only render the link if the site settings have been loaded.
* Removes `isRequired` from the siteId prop type in `JetpackModuleToggle` definition.

#### Testing instructions

1. Checkout this branch
1. Clear the redux state cache
1. Go to `/settings/general/:site`, for one of your sites.
1. Check the page loads correctly and no propTypes warnings and shown in the console.
1. Clear the redux state cache
1. Go to `/settings/writing/:site`, for one of your sites.
1. Check the page loads correctly and no propTypes warnings and shown in the console.
1. Clear the redux state cache
1. Go to `/settings/discussion/:site`, for one of your sites.
1. Check the page loads correctly and no propTypes warnings and shown in the console.
1. Clear the redux state cache
1. Go to `/settings/traffic/:site`, for one of your sites.
1. Check the page loads correctly and no propTypes warnings and shown in the console.
1. Go to `/settings/`, pick a site
1. Verify the general settings page loads correctly.
1. Clean your Redux state & cache and test the previous steps again.

### CODE REVIEW

File: client/my-sites/site-settings/controller.js [hunk 1]
```diff
@@ -111,38 +99,29 @@ module.exports = {
...
 	guidedTransfer( context ) {
 		renderPage(
 			context,
-			<SiteSettingsComponent sites={ sites } section="guidedTransfer" hostSlug={ context.params.host_slug } />
+			<SiteSettingsComponent section="guidedTransfer" hostSlug={ context.params.host_slug } />
 		);
 	},
 
 	deleteSite( context ) {
-		let site = sites.getSelectedSite();
```
Comment (tyxla) [inline client/my-sites/site-settings/controller.js]: Before getting rid of this in `<DeleteSite`, it has to be refactored, because it will fail when deleting a site. Currently, it uses the sites list actions to delete the site, and if you pass a Redux site object, it'll not remove the site from sites-list, creating odd behavior. So I suggest tackling that first, and then migrating this one.

**Comment by oskosk** [comment]
@tyxla I've labeled this as in progress because I've found a few errors that extend to the monitor settings component

**Comment by oskosk** [comment]
This is ready for another review @tyxla.

File: client/my-sites/site-settings/controller.js (hunk 1 above)
Comment (oskosk) [inline client/my-sites/site-settings/controller.js]: I've removed everything related to the deleteSite route from this PR as it was here previously but belonged to another branch.

File: client/my-sites/site-settings/controller.js [hunk 2]
```diff
@@ -64,8 +64,7 @@ module.exports = {
 	siteSettings( context ) {
 		let analyticsPageTitle = 'Site Settings';
 		const basePath = route.sectionify( context.path );
-		const fiveMinutes = 5 * 60 * 1000;
-		let site = sites.getSelectedSite();
+		const site = getSelectedSite( context.store.getState() );
```
Comment (tyxla) [inline client/my-sites/site-settings/controller.js]: WIth a clean redux state this will break the `site.*` calls below. We should either guard against these calls if `site` is `null` (by specifically prepending a `site` check), or use selectors for all of those calls.

Comment (oskosk) [inline client/my-sites/site-settings/controller.js]: I'll revisit this more exhaustively because this error is happening in other components too. Look like I was not clearing the cache properly when testing this

Comment (oskosk) [inline client/my-sites/site-settings/controller.js]: This has been addressed in this PR

File: client/my-sites/site-settings/section-security.jsx [hunk 3]
```diff
@@ -24,7 +24,7 @@ const SiteSettingsSecurity = ( { site, translate } ) => {
 		);
 	}
 
-	if ( ! site.canManage() ) {
+	if ( ! site.canManage ) {
```
Comment (oskosk) [inline client/my-sites/site-settings/section-security.jsx]: This has also been addressed in this PR

**Comment by oskosk** [comment]
@tyxla this is ready for another review. If it's too big I can split it in two or three others.

File: client/my-sites/site-settings/main.jsx [hunk 4]
```diff
@@ -90,7 +78,7 @@ export class SiteSettingsComponent extends Component {
 	}
 
 	render() {
-		const { site } = this.state;
+		const { site } = this.props;
```
Comment (tyxla) [inline client/my-sites/site-settings/main.jsx]: `site` is passed as a connected prop for `<SiteSettingsNavigation />`, so it's no longer necessary to be specifically passed to it, you can remove it.

Also, I think after you remove it from here, `site` will no longer need to be added as a connected prop, so it can be removed from there, too. You can use `siteId` for the checks in `render()`.

File: client/my-sites/site-settings/controller.js [hunk 5]
```diff
@@ -78,27 +77,18 @@ module.exports = {
...
-			} else {
-				sites.once( 'change', function() {
-					site = sites.getSelectedSite();
-					site.fetchSettings();
-				} );
-			}
-		}
-
-		const upgradeToBusiness = () => page( '/checkout/' + site.domain + '/business' );
+		const upgradeToBusiness = site
```
Comment (tyxla) [inline client/my-sites/site-settings/controller.js]: `upgradeToBusiness` appears to be no longer used here - it's used only in Traffic, so this can be completely removed.

File: client/my-sites/site-settings/main.jsx [hunk 6]
```diff
@@ -25,31 +24,21 @@ import SiteSettingsNavigation from './navigation';
...
-			site: this.props.sites.getSelectedSite()
-		};
-	}
 
-	componentWillMount() {
-		debug( 'Mounting SiteSettings React component.' );
-		this.props.sites.on( 'change', this.updateSite );
-	}
+	static propTypes = {
+		hasLoadedSitePurchasesFromServer: PropTypes.bool.isRequired,
```
Comment (tyxla) [inline client/my-sites/site-settings/main.jsx]: Can we separate the connected props from the rest of them? Something I've seen being done recently:

```
siteId: PropTypes.number,
// Connected props
isRequesting: PropTypes.bool.isRequired,
```

File: client/my-sites/site-settings/main.jsx [hunk 7]
```diff
@@ -25,31 +24,21 @@ import SiteSettingsNavigation from './navigation';
...
+		site: PropTypes.object,
+		siteId: PropTypes.number,
+		jetpackSettingsUiSupported: PropTypes.bool
+	};
 
-	componentWillUnmount() {
-		this.props.sites.off( 'change', this.updateSite );
-	}
+	static defaultProps = {
+		section: 'general'
```
Comment (tyxla) [inline client/my-sites/site-settings/main.jsx]: Does it make sense to add default props for all non-required props?

File: client/my-sites/site-settings/navigation.jsx [hunk 8]
```diff
@@ -2,37 +2,38 @@
...
+import { getSelectedSite } from 'state/ui/selectors';
 
-export default React.createClass( {
-	displayName: 'SiteSettingsNavigation',
+export class SiteSettingsNavigation extends React.Component {
 
-	propTypes: {
-		site: React.PropTypes.object.isRequired,
+	static propTypes = {
+		site: React.PropTypes.object,
```
Comment (tyxla) [inline client/my-sites/site-settings/navigation.jsx]: Can we separate the connected props from the ones we expect to be passed specifically?

File: client/my-sites/site-settings/navigation.jsx [hunk 9]
```diff
@@ -2,37 +2,38 @@
...
  */
 import config from 'config';
 import SectionNav from 'components/section-nav';
 import NavTabs from 'components/section-nav/tabs';
 import NavItem from 'components/section-nav/item';
+import { getSelectedSite } from 'state/ui/selectors';
 
-export default React.createClass( {
-	displayName: 'SiteSettingsNavigation',
+export class SiteSettingsNavigation extends React.Component {
```
Comment (tyxla) [inline client/my-sites/site-settings/navigation.jsx]: Maybe import `{ Component } from 'React'` specifically, so you don't have to say `React.Component` but just `Component`.

File: client/my-sites/site-settings/press-this/index.jsx [hunk 10]
```diff
@@ -50,13 +50,17 @@ class PressThis extends Component {
 						) }
 					</p>
 					<p className="pressthis">
```
Comment (tyxla) [inline client/my-sites/site-settings/press-this/index.jsx]: Maybe it would make sense to not display the `<p>` if the site is `null` - currently, if there's no `site`, an empty `<p>` will be displayed.

File: client/my-sites/site-settings/controller.js [hunk 11]
```diff
@@ -78,27 +77,18 @@ module.exports = {
 		}
 
 		// if user went directly to jetpack settings page, redirect
-		if ( site.jetpack && ! config.isEnabled( 'manage/jetpack' ) ) {
+		if ( site && site.jetpack && ! config.isEnabled( 'manage/jetpack' ) ) {
```
Comment (tyxla) [inline client/my-sites/site-settings/controller.js]: Maybe use `isJetpackSite` and `getSelectedSiteId` here? Once we get rid of the `upgradeToBusiness` below, we will not need the entire `site` object and we can use more specific selectors. Also, you can use `canCurrentUser( state, siteId, 'manage_options' )` instead of `utils.userCan` that needs a `site` object.

File: client/my-sites/site-settings/jetpack-module-toggle.jsx [hunk 12]
```diff
@@ -31,7 +31,7 @@ class JetpackModuleToggle extends Component {
 	};
 
 	static propTypes = {
-		siteId: PropTypes.number.isRequired,
```
Comment (tyxla) [inline client/my-sites/site-settings/jetpack-module-toggle.jsx]: What is the rationale behind this change? I think this component does not function properly without a `siteId, so I'd suggest keeping this one required, but instead, altering the logic either in this component or in the parent component to not render the toggle if `siteId` is `null`. Or is there other specific reason to make it non-required?

File: client/my-sites/site-settings/main.jsx [hunk 13]
```diff
@@ -25,31 +24,21 @@ import SiteSettingsNavigation from './navigation';
...
-	constructor( props ) {
-		super( props );
-
-		// bound methods
-		this.updateSite = this.updateSite.bind( this );
-
-		this.state = {
-			site: this.props.sites.getSelectedSite()
-		};
-	}
```
Comment (tyxla) [inline client/my-sites/site-settings/main.jsx]: Not sure why our linter does not yell about trailing empty lines in classes. Care to remove this one?

File: client/my-sites/site-settings/main.jsx [hunk 14]
```diff
@@ -25,31 +24,21 @@ import SiteSettingsNavigation from './navigation';
...
-	componentWillMount() {
-		debug( 'Mounting SiteSettings React component.' );
-		this.props.sites.on( 'change', this.updateSite );
-	}
+	static propTypes = {
+		hasLoadedSitePurchasesFromServer: PropTypes.bool.isRequired,
+		purchasesError: PropTypes.object,
+		section: PropTypes.string,
+		sitePurchases: PropTypes.array.isRequired,
+		site: PropTypes.object,
```
Comment (tyxla) [inline client/my-sites/site-settings/main.jsx]: If you follow my suggestions for removing the `site` connected prop, don't forget to remove it from here, too.

File: client/my-sites/site-settings/main.jsx [hunk 15]
```diff
@@ -67,7 +56,6 @@ export class SiteSettingsComponent extends Component {
 	}
```
Comment (tyxla) [inline client/my-sites/site-settings/main.jsx]: Just noticed we're not even using the `getStrings()`. Care to clean it up? It would also rid us of importing`i18n`.

File: client/my-sites/site-settings/section-security.jsx [hunk 16]
```diff
@@ -24,7 +26,7 @@ const SiteSettingsSecurity = ( { site, translate } ) => {
...
```
Comment (tyxla) [inline client/my-sites/site-settings/section-security.jsx]: How about using the `isJetpackSite` selector instead of `site.jetpack` here? The others should be fine, as they're computed and will be there if `.jetpack` is set.

File: client/my-sites/site-settings/controller.js [hunk 17]
```diff
@@ -78,27 +77,18 @@ module.exports = {
...
 			window.location.href = '//wordpress.com/manage/' + site.ID;
 			return;
 		}
 
-		if ( ! site.latestSettings || new Date().getTime() - site.latestSettings > ( fiveMinutes ) ) {
-			if ( sites.initialized ) {
-				site.fetchSettings();
```
Comment (tyxla) [inline client/my-sites/site-settings/controller.js]: One thing that I almost missed here: if we're removing the code that fetches the site settings, let's make sure we're querying settings in the affected components (that might require adding a `<QuerySiteSettings />` here and there).

File: client/my-sites/site-settings/jetpack-module-toggle.jsx (hunk 12 above)
Comment (oskosk) [inline client/my-sites/site-settings/jetpack-module-toggle.jsx]: Well, the rationale was that it was required when we thought we would never be handling a `null` value here. It was there before the usage of the redux site selectors was widespread among the settings tabs and a lot more render blocking techniques were used before reaching the JetpackModuleToggle rendering.

Looks to me that if we're making the whole system be able to handle a `null` value for sites or site ids, we should make this toggle component at least be able to render without warnings when that particular state is reached by the system. Also this non required usage of a siteId prop looks to be the case for 80% of the components that use a `siteId` as prop.

Otherwise, if we follow techniques like `{ siteId && <JetpackModuleToggle siteId={siteId} /> }` we would probably need to do that everywhere for consistency even when we don't detect a warning right now. For example, here in this case, it will demand that 3 parent components make that check. And that's just one of the tabs. 

What is stopping us from currently making every child component on every piece of Calypso be conditionally rendered if the props we're passing to it are null or such? This last rationale is what I was trying to avoid with this change.

Comment (oskosk) [inline client/my-sites/site-settings/jetpack-module-toggle.jsx]: Another alternative that came to my mind was making the JetpackModuleToggle don't depend on a passed down prop, and have it as a connected prop instead based on `getSelectedSiteId`. But this would also mean that for a render without warnings, the prop should be declared as non required.

Comment (tyxla) [inline client/my-sites/site-settings/jetpack-module-toggle.jsx]: Thanks for the clarification. Let's leave it as you suggest and we can always go back to it if necessary.

File: client/my-sites/site-settings/main.jsx (hunk 15 above)
Comment (oskosk) [inline client/my-sites/site-settings/main.jsx]: done!

File: client/my-sites/site-settings/main.jsx (hunk 13 above)
Comment (tyxla) [inline client/my-sites/site-settings/main.jsx]: It's still there 😉

File: client/my-sites/site-settings/section-security.jsx (hunk 16 above)
Comment (tyxla) [inline client/my-sites/site-settings/section-security.jsx]: I recommended this one, because `cannot get .jetpack of undefined` is one of the often errors reported in Calypso, and by protecting with the selector we can help mitigate such errors.

Comment (oskosk) [inline client/my-sites/site-settings/section-security.jsx]: done!, thanks!
